### ISSUE TITLE
Updating child via child.Update() does not work

### ISSUE THREAD

**Issue by budachst** [starkillerOG/motion-blinds#7]
I noticed that I cannot use .Update() but only .Update_from_cache? Is this maybe something between my gateway and my blind? Happens that I also can't seem to pick-up any multicast transmissions from my gateway. Is this maybe specific to a certain version of HW/SW regarding the gateway?

**Comment by starkillerOG** [comment]
.Update() uses direct communication with the blind and relays on udp multicast. For it to work the device you run .Update() from and the motion_blinds gateway need to be in the same VLAN, furthermore udp multicast needs to be allowed by your router.

.Update_from_cache() takes the status from the gateway and does not need udp mulitcast, it will return the latest states the gateway has in memory (since the last time the gateway received status from the blind), status updates happen when you open the motion blinds app, move the blind (in any way) or you call the .Update() function.

**Comment by budachst** [comment]
The communication is running via the gateway, isn't it? Its not, that the blinds themselves are on the WLAN as well, or are they? When running tcpdump and sniffing the multicast networks, I can see quite a lot of multicast packets, but none from or to the gateway. Besides, probably having a network issue, which I would have to investigate, but I do seem a lot of multicast traffic on my LAN/WLAN - what's the main advantage of running blind.Update()? Will it provide any other data than what the gateway has? I seem to remember, that blind sends out a multicast push after it has finished its current action, so the advantanges of multicast seem to be rather limited.

**Comment by starkillerOG** [comment]
yes the communication runs through the gateway.
you are correct the blinds are not on your network (they use 433 MHz RF communication).

blind.Update() will get the current status, so for example while you are charging your blind, blind.Update() will give you the actual battery values and you can see the battery charging in time. However blind.Update_from_cache() will continue reporting the old battery value untill you open the motion blinds app or move the blind (you can't track the charging of the battery). Same thing holds for the signal_strength.

Another example: while a blind is moving blind.Update_from_cache() will return the previous position when the blind was still stationarry (until the blind completes the movement), blind.Update() will give you the actual current position as it is moving and you can see the postion changing untill it reaches its target position.

**Comment by starkillerOG** [comment]
Another example: if the gateway misses the 433MHz push message from a blind at the end of a movement (because of low signal strenght), the gateway will continue to give the incorrect old position value (using the blind.Update_from_cache()), you will not be able to detect that the position is actually wrong using the blind.Update_from_cache(). However using the blind.Update() you will get a fresh status update so you are garantueed that the values reported are actually correct. If the communication fials, you will get a timeout exception from the blind.Update() function, so you know that your values might be incorrect.

**Comment by budachst** [comment]
I see - thanks. I spent quite a good amount of time yesterday to check my network, but I couldn't come up with a possible reason, as of why multicast should not work. I am seeing multicast traffic on my WLAN/LAN - also on the host on which I am running the Python app, but I do never see any multicasts from the motion gateway.

As far as I understood, the gateway will always send a multicast push, when a blind has finished it's operation, so even when not requesting a update via .Update() there should be multicasts on 238.0.0.18, right?

**Comment by starkillerOG** [comment]
yes you are right, at the end of a movement there schould be a multicast visable on '238.0.0.18'.
Note that I have experianced problems when the gateway was connected through an acces point that is connected to my router on the same VLAN.
So it might help to move the gateway and the device where you are sniffing from both to your main router (not through an acces point or something).

**Comment by budachst** [comment]
Well, since the gateway is Wifi only, the gateway has to be on it and run through an access point. However, I also have tried to move the gateway to another WLAN (I do have two) and the node sniffing is on the LAN anyway. I really can't imagine, that the gateway would not work on either one of my two APs, but even that cannot be reliably said. Maybe I'll spin up an ad-hoc WLAN on my Mac and hook the gateway up there.

FWIW, I do see multicats on 238.0.0.18 from another device on my network, which is a IPTV which is hooked up to my satellite dish, but other than that, there's absolutely no other traffic on 238.0.0.18.

**Comment by d-EScape** [comment]
I'm using the module in a domoticz plugin and encountered the same problem. The state of the blinds is accurately returned on every Update_from_cache(), but the battery voltage is not, so I am still trying to get .Update() to work. It fails with the same time-out, but the battery_voltage is updated afterwards! So something seems to get trough(?)

The bridge is on WiFi (obviously ;-) and the client (normally) on ethernet. I have also tried an old USB Wifi adapter on the client, so the client was on the same 2,4GHz wlan, but still got the time-out. So communicating between WiFi and Ethernet does not seem te be the problem.

**Comment by budachst** [comment]
Well, I do get updates for the battery voltage via Update_from_cache(), so I don't have that issue. However, I sniffed my complete network for multucast traffic on 238.0.0.18 but other than my NetstreamSat, which is  s Sat->IP device, nothing seems to happening otherwise on that multicast address.

**Comment by d-EScape** [comment]
I dit also get a battery value using Update_from_cache() but it was always the same and it changed immediately with Update(). Still got the timed-out error though. 

I spend most of the day researching this issue, but i can't figure it out. Using  `tcpdump -ni <interface> igmp` I tried to pin down where the multicast (238.0.0.18) gets lost. On my Asus ac86u (in a mesh setup) I could see the igmp traffic on:
br0 (ethernet/lan bridge?) but NOT on eth5 (which is the physical 2,4 wifi). However the traffic IS visible on eth5.0 (which I _think_ was created Asus aimesh)

I would nog be surprised if this had something to do with the mesh setup, but that would not explain why the traffic is visible on br0. 

On the client side I tried to manually join the multicast using: `sudo ip maddr add 238.0.0.18`. But it seemed to not be saved, so i am missing something. Setting the interface of the client to "ALLMULTI' did not fix anything. But I do think that it would make the multicast show up in tcpdump even without joining (I am not sure about this!). However I still do not see the igmp tracffic on the client.

About the battery value getting updated: When running Update() I also see normal traffic (not just multicast) from and to the bridge. To me it looks like the real data is send as normal traffic and the multicast is used mainly for some kind of handshaking/control? I am by no means an expert so it's more guesswork than knowledge at this point :-$

**Comment by starkillerOG** [comment]
> 
> 
> I'm using the module in a domoticz plugin and encountered the same problem. The state of the blinds is accurately returned on every Update_from_cache(), but the battery voltage is not, so I am still trying to get .Update() to work. It fails with the same time-out, but the battery_voltage is updated afterwards! So something seems to get trough(?)
> 
> The bridge is on WiFi (obviously ;-) and the client (normally) on ethernet. I have also tried an old USB Wifi adapter on the client, so the client was on the same 2,4GHz wlan, but still got the time-out. So communicating between WiFi and Ethernet does not seem te be the problem.
> 
> Edit: Unsurprising I can't get the MotionMulticast functionality to trigger handler functions either. So does this mean there is simply no multicast traffic getting through? But if so, how did it manage to update the battery status anyway?

So this means the multicast traffic is just not getting through.

The reason why the battery status is updated:
When issuing a `.Update()` the gateway is instructed to communicate with the blind to get the latest info (that happens over direct udp communication, so that message arrives at the gateway, if that were to fail the error message would be diffrent, so the acknoladgement of this instruction is received).
The gateway then communicates with the blind over 433MHz RF and succesfully obtains the new status including the new battery value of the blind, at this point the internal cache of the gatway is updated so a succesif call of `Update_from_cache()` will give you the new status including the new battery value. Now the gateway sends out a udp multicast message containing the new status. This is actually where it goes wrong, somehow this udp multicast message is beeing blocked in your network and does not arrive at your client. Then you get the timeout error.

**Comment by starkillerOG** [comment]
To go into some more detail:
After issuing the `.Update()` over direct udp communication the gateway acknoladges using direct udp communciation to the client with the status of the blind from cache, this status is processed and does update all values within the python class, but since this is from cache, it is the status obtained at the previous communication with the blind (`.Update()`). So basically that status is always lacking behind 1 `.Update()` call. The python `.Update()` call will then wait untill the multicast push is beeing received that contains the actual new status of that `.Update()` call (to not be lacking behind). That is what fails for you.

So basically if you keep calling `.Update()` and just ignore the errors you will get new status of the blind, but each `.Update()` call will lack behind 1 call, you basically are getting the status of the privious call. (This is assuming there is enough time in between the `.Update()` calls to let the gateway finish it's communication with the blind, withouth the udp multicast response, there is no way of knowing when the gateway finishes it's communication with the blind and when the status is updated in the cache).

**Comment by d-EScape** [comment]
"So basically if you keep calling .Update() and just ignore the errors you will get new status of the blind, but each .Update() call will lack behind 1 call, you basically are getting the status of the privious call. " 
That makes sense.

The Update() method was probably never intended as a synchronous call. Maybe you could add a async_Update() or trigger_Update() or something like that, that doesn't wait and does nog generate errors?

I have done some more testing. Dug up an old Cisco router running Tomato firmware and connected the bridge and (wired) client to it. De Update() works like a charm in that setup. Unfortunately I am not planning to go back to a single router Wifi-N setup ;-) At least it shows that the Asus and/or iaMesh setup is not handling the multicast as expected. I'll report back if I ever find a solution.

Anyhow: thanks for your input and for the great module!

**Comment by budachst** [comment]
Ok, after I looked at the pcap using Wireshark, I am wondering, why my node specifically seem to no want to be included in the bridge's multicast traffic:

```
Internet Group Management Protocol
    [IGMP Version: 3]
    Type: Membership Report (0x22)
    Reserved: 00
    Checksum: 0x0aef [correct]
    [Checksum Status: Good]
    Reserved: 0000
    Num Group Records: 2
    Group Record : 238.0.0.18  Mode Is Exclude
        Record Type: Mode Is Exclude (2)
        Aux Data Len: 0
        Num Src: 0
        Multicast Address: 238.0.0.18
```

Maybe this is the issue I am having.

**Comment by d-EScape** [comment]
The same might be happening in my setup. What settings did you use to get this output? I am using Tshark on a headless linux machine and am trying to figure out a useful method for capturing this igmp interaction. 

I'm getting:
    6	5.027994308	192.168.19.222	→	238.0.0.18  	IGMPv2	48	Membership Report group 238.0.0.18
    7	5.691877593	192.168.19.1	→	224.0.0.1   	IGMPv2	62	Membership Query, general
    8	8.580009809	192.168.19.222	→	238.0.0.18  	IGMPv2	48	Membership Report group 238.0.0.18
    9	9.603997961	192.168.19.96	→	239.255.255.250	IGMPv2	48	Membership Report group 239.255.255.250
   10	10.047358634	192.168.19.222	→	224.0.0.2   	IGMPv2	48	Leave Group 238.0.0.18

9 is probably not related to this issue. I don't know if 7 is. But 10 shows the "leave group". It might be the same "Exclude"?

I'm nu networking expert, but maybe the client doesn't know how to ask the router for directions (routing issue on the client?)
Or the router doesn't know what to do with the request (routing/propagation issue in the router?) 

Because it does work on another router I suspect the routing issue on the router. What setup are you using? Mine is a Asus ac86u combined with ac68u in an iaMesh.

**Comment by d-EScape** [comment]
Found it! At least for my setup.
After disabling IGMP snooping on the 2.4 wireless interface of my Asus router the IGMP packets get through and the .Update() works as expected.

Learned something today. IGMP snooping (which is enabled by default on my router) does not enable IGMP traffic, but "optimizes" it by analyzing it. Not as intelligently as we would like, as it seems ;-)

**Comment by budachst** [comment]
Would you mind sharing some lines of traffic, so I can see, how it would show up in tcpdump or wireshark? I had scouted my Orbi Wifi mesh nodes for igmp snooping, but could only find igmp proxy, which was not enabled on either one.

**Comment by d-EScape** [comment]
I don't mind at all, but I am not sure if all traffic is related to the .Update() command being run. 

   13 24.117106405 192.168.19.222 → 238.0.0.18   IGMPv2 48 Membership Report group 238.0.0.18
   14 27.701099550 192.168.19.222 → 224.0.0.251  IGMPv2 48 Membership Report group 224.0.0.251
   15 29.910929163 192.168.19.1 → 224.1.1.50   IGMPv2 62 Membership Query, specific for group 224.1.1.50
   16 30.930051854 192.168.19.1 → 224.1.1.50   IGMPv2 62 Membership Query, specific for group 224.1.1.50
   17 40.040067054 192.168.19.1 → 224.0.0.1    IGMPv2 62 Membership Query, general
   18 42.293113073 192.168.19.222 → 238.0.0.18   IGMPv2 48 Membership Report group 238.0.0.18
   19 42.805094632 192.168.19.222 → 239.255.255.250 IGMPv2 48 Membership Report group 239.255.255.250
   20 46.133068620 192.168.19.222 → 224.0.0.251  IGMPv2 48 Membership Report group 224.0.0.251
   21 48.212500019 192.168.19.1 → 224.1.1.50   IGMPv2 62 Membership Query, specific for group 224.1.1.50
   22 49.230136541 192.168.19.1 → 224.1.1.50   IGMPv2 62 Membership Query, specific for group 224.1.1.50

I am using tshark or tcpdump, but am pretty new at it. If you give me a command (filter etc) to run then I will run it and share the output.

**Comment by budachst** [comment]
Hmm… that's what I wanted to know myself. ;) I checked the interface of my client and it is actually subscribed to the multicast group 238.0.0.18, as both ip madr and netstat -g show this:

root@fhem:~# ip maddr show
1:	lo
	inet  224.0.0.1
	inet6 ff02::1
	inet6 ff01::1
2:	ens18
	link  33:33:00:00:00:01
	link  01:00:5e:00:00:01
	link  01:00:5e:00:00:fb
	link  33:33:ff:55:c6:00
	link  33:33:00:00:00:fb
	link  01:00:5e:00:00:12
	inet  238.0.0.18
	inet  224.0.0.251
	inet  224.0.0.1
	inet6 ff02::fb
	inet6 ff02::1:ff55:c600 users 2
	inet6 ff02::1
	inet6 ff01::1
root@fhem:~# netstat -g
IPv6/IPv4-memberships
Schnittstelle   RefZäh Grupp
--------------- ------ ---------------------
lo              1      all-systems.mcast.net
ens18           1      238.0.0.18
ens18           1      224.0.0.251
ens18           1      all-systems.mcast.net
lo              1      ip6-allnodes
lo              1      ff01::1
ens18           1      ff02::fb
ens18           2      ff02::1:ff55:c600
ens18           1      ip6-allnodes
ens18           1      ff01::1

So, since I am receiving multicast traffic on 224.0.0.251  and since ens18 is a member of 238.0.0.18, it should receive these messages as well.

Looking at your tcpdump, I also never saw anything else related to 238.0.0.18 as the packets you are seeing.

**Comment by starkillerOG** [comment]
> The Update() method was probably never intended as a synchronous call. Maybe you could add a async_Update() or trigger_Update() or something like that, that doesn't wait and does nog generate errors?

Yes I will do that if that is conveniant for you

**Comment by starkillerOG** [comment]
> Found it! At least for my setup.
> After disabling IGMP snooping on the 2.4 wireless interface of my Asus router the IGMP packets get through and the .Update() works as expected.

I will add this to the HomeAssistant documentation as a troubleshooting hint

**Comment by budachst** [comment]
Great addiditon. I tried to incorporate that function into my little project and it works very well. You'll have to be a bot careful though, since calling blind.Update_trigger() too often, will bog down the radio connection between the blinds and the gateway. Either the gateway has some kind of enforcement on the wireless 433Mhz channel, or the connection to the blinds itself is rather slow and can be easily overwhelmed. Which would make sense, since no one would implement multicast updates without a reason, I guess.

So, I am using this function cautiously for the time being and checking, what the minimum interval is, when running 3 blinds.

**Comment by d-EScape** [comment]
I am forcing an update only once a day and even that should be more than enough. The blinds data in the bridge cache is updated on every action (opening/closing the blinds), in which case Update_from_cache will already return the correct values. Forcing updates is only needed if the blinds where not operated for a long time and you still want an actual battery level reading.
Frequently communicating to the bridge will probably also drain the battery faster.
It goes without saying that all of this only affects the battery powered version of the motor ;-) Otherwise you would not need the forced update at all, since the bridge always knows the position of the blinds anyway. Even when they where operated from another remote.

**Comment by starkillerOG** [comment]
yes you are right that it may drain the battery of the blind a bit faster, although I would guess that moving the blind once will drain a lot more battery than only 433MHz communication. Would be very intresting to know how many communications equal one time moving the blind in regards to the battery drain.

For HomeAssistant an update is called every 10 minutes and multicast pushes are processed imediatly (when a blind moves).
If a blind becomes unavailable the update ferquency will increase to once every 1 minute in order to let the blind be available as soon as possible.

**Comment by d-EScape** [comment]
Interesting ... we are getting a bit of topic, but I'm curious. What it the advantage of running a (real) update instead of the update_from_cache in your use case?
You are absolutely right about the battery drain of communication vs operating the blinds only once, but still... over a longer time it might be noticeable. Just like the amount of electronics in modern cars (alarm/remote assistance etc). They just sip battery power, but if you leave your car in the garage for a couple of months the battery will drain (completely).

**Comment by budachst** [comment]
What I am more curious about is actually why I seem to be able to crash my motion bridge, when issuing too many of those Update.trigger() commands on a regular basis. I do agree that it may enough to do that once a day, when the blinds are not operated, but nevertheless, the bridge shouldn't crash.

Note, that can't say for sure if it's actually crashing, but at least the blue LED turns red when you request too many of those updates. If it's only my bridge, than it might be faulty…

**Comment by d-EScape** [comment]
If it works anything like common ('Klik aan klik uit' in the Netherlands) 433 switchable power outlets then it will shoutout every signal several times, just to make sure it is actually received. That should not be necessary with true(!) bi-directional communication, but I wouldn't be surprised if it is just that inefficient. The bridge shouting out an update request ten times and every blind responding ten times. Sending the same request in short intervals would make it increasingly hard to distinguish the individual signals. But again: that is just a guess/me thinking out loud.

**Comment by starkillerOG** [comment]
I will close this issue since I think all problems are resolved.
Feel free to open a new issue if you have more questions/problems
